Resolve #299 - Untargettable buildings

Kubeer opened the pull request:
*Buildings now can be destroyed only in the correct order
*Untargettable buildings also have 99999 armor and magic resistance

Refs #299

Kubeer committed:
Resolve #299 - Untargettable buildings

*Buildings now can be destroyed only in the correct order
*Untargettable buildings also have 99999 armor and magic resistance

Refs #299

Kubeer committed:
Standards matching

danil179 commented:
Why do we need the protection if the buildings are untargettable?

danil179 commented on GameServerLib/GameObjects/AttackableUnits/AI/BaseTurret.cs:
```diff
@@ -94,6 +137,39 @@ public override void Update(float diff)
                 _game.PacketNotifier.NotifySetTarget(this, null);
             }
 
+            if (_dependOnAll != null || _dependOnSingle != null)
```
All this part can go to a function `UpdateProtection`.

Kubeer commented:
In issue report I saw that they have to be invulnerable so just to make sure I added this.

danil179 commented:
> 
> 
> In issue report I saw that they have to be invulnerable so just to make sure I added this.

But is it essential? If not, please remove this part as it clutters the code and complicate things.

Kubeer committed:
Removed stats protection and moved logic to an UpdateProtection function

danil179 commented:
I think that this solution is not quite good. You should add something like "ProtectionManager" instance to each building that will get and handle all of these protection mechanisms and will avoid the repetitive code and the too much arguments in the constructors. I also  was thinking about doing class "ProtectedBuilding", but this will create a lot of problems so please don't try to solve it with inheritances.

lzardy commented:
> I think that this solution is not quite good. You should add something like "ProtectionManager" instance to each building that will get and handle all of these protection mechanisms and will avoid the repetitive code and the too much arguments in the constructors. I also was thinking about doing class "ProtectedBuilding", but this will create a lot of problems so please don't try to solve it with inheritances.

Having a central ProtectionManager would also be useful for invulnerability for disconnected players.

Kubeer commented:
Ok, so I'll make a ProtectionManager but should i commit it here or make a new branch instead?

lzardy commented:
> Ok, so I'll make a ProtectionManager but should i commit it here or make a new branch instead?

Here. You can make a new PR for replacing the current invuln. for disconnected players with ProtectionManager.

xxx3456789 commented:
if disconnect then apply invulnerability 
and edit something in Damage or Buff be like
if this object have invulnerability, then "TakeDamage" 's damage value will be 0 (its conception so only take for example)

Kubeer committed:
Replaced logic with ProtectionManager

Logic for invulnerability for disconnected players will be added in next PR since it'd get messy here.

Kubeer commented:
Maybe some functions needs renaming but I couldn't came up with something better than this.

danil179 reviewed (approved):
Approved.
